From 363361a9e26322d9e6a87e17effc50add8653533 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Fri, 7 Mar 2014 19:31:42 +0000
Subject: [PATCH 24/56] Fix 0b9470c Fix fileringbuffer locking...

Commit 0b9470c (Fix fileringbuffer locking up on read()) modified
FileRingBuffer::safe_read to use fstat() to check the file size
before reading.  The new code compares (internalreadpos + tot)
with the reported size.  This is mostly correct except when
MythPlayer::SwitchProgram detects the transition across a livetv
programme boundary.  In this case internalreadpos holds the read
position for the old file until the decoder calls FileChanged().

This patch includes readAdjust (set by SetAdjustFilesize()) to
correctly determine the current read position in the file.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
Signed-off-by: Jean-Yves Avenard <jyavenard@mythtv.org>
---
 mythtv/libs/libmythtv/fileringbuffer.cpp |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/mythtv/libs/libmythtv/fileringbuffer.cpp b/mythtv/libs/libmythtv/fileringbuffer.cpp
index a092533..b3996c8 100644
--- a/mythtv/libs/libmythtv/fileringbuffer.cpp
+++ b/mythtv/libs/libmythtv/fileringbuffer.cpp
@@ -476,7 +476,8 @@ int FileRingBuffer::safe_read(int fd, void *data, uint sz)
         ret = fstat(fd2, &sb);
         if (ret == 0 && S_ISREG(sb.st_mode))
         {
-            if ((internalreadpos + tot) >= sb.st_size)
+            long long llpos = (internalreadpos - readAdjust) + tot;
+            if (llpos >= sb.st_size)
             {
                 // We're at the end, don't attempt to read
                 read_ok = false;
@@ -486,7 +487,7 @@ int FileRingBuffer::safe_read(int fd, void *data, uint sz)
             else
             {
                 toread  =
-                    min(sb.st_size - (internalreadpos + tot), (long long)toread);
+                    min(sb.st_size - llpos, (long long)toread);
                 if (toread < (sz-tot))
                 {
                     eof = true;
-- 
1.7.10.2

